charms/knative-operator/src/charm.py

```diff
@@ -142,6 +186,29 @@ def _on_knative_operator_pebble_ready(self, event):
         self.unit.status = MaintenanceStatus("Configuring Pebble layer")
         self._update_layer(event)
 
+    def _on_otel_collector_relation_changed(self, event):
+        """Event handler for on['otel-collector'].relation_changed."""
+        # Apply all changes only if the otel collector has not been deployed
+        if not self._otel_exporter_ip:
+            self.unit.status = MaintenanceStatus("Applying otel collector")
+            self._template_files_ext = [
+                f"{self._src_dir}/manifests/observability/collector.yaml.j2"
+            ]
+            # Reset the resource handler so it uses the expanded template files list
+            self._resource_handler = None
+            self._apply_all_resources()
+        # Write to its own application bucket
+        relation_data = self.model.get_relation("otel-collector", event.relation.id).data[self.app]
+        # Update own application bucket with otel collector information
+        relation_data.update(
+            {
+                "otel_collector_svc_namespace": self.model.name,
+                "otel_collector_svc_name": "otel-collector",
+                "otel_collector_port": "55678",
+            }
+        )
```

Comment:
Can we add some unit tests that ensure this works as expected?

Reply:
@DnPlas I think this is still outstanding - we do have a test that checks whether the relation-created event causing the handler to fire, but we do not have a test that asserts the data that should get into the relation actually makes it there (eg: we know the charm *tries* to handle a relation created, but not that it does it successfully)

Reply:
I have added a test case for it. Let me know if that suffices.

Reply:
lgtm!